File: androidDemo/src/main/java/com/novage/p2pml/demo/ui/screens/player/components/VideoPlayerSurface.kt
package com.novage.p2pml.demo.ui.screens.player.components

import android.annotation.SuppressLint
import android.view.LayoutInflater
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.media3.common.Player
import androidx.media3.common.util.UnstableApi
import androidx.media3.ui.PlayerView
import com.novage.p2pml.demo.R

private const val ASPECT_RATIO = 16f / 9f

@SuppressLint("InflateParams")
@androidx.annotation.OptIn(UnstableApi::class)
@Composable
fun VideoPlayerSurface(player: Player?, isP2PActive: Boolean, isVideoReady: Boolean, onSettingsClick: () -> Unit) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .aspectRatio(ASPECT_RATIO)
            .background(Color.Black)
    ) {
        if (player != null) {
            AndroidView(
                factory = { ctx ->
                    // Inflate XML to force TextureView, resolving Z-order glitches during Compose animations
                    val view = LayoutInflater.from(ctx)
                        .inflate(R.layout.layout_player_view, null) as PlayerView

                    view.apply {
                        useController = true
                        setShowBuffering(PlayerView.SHOW_BUFFERING_ALWAYS)
                    }
                },
                update = { playerView ->
                    playerView.player = player
                },
                modifier = Modifier.fillMaxSize()
            )
        }

        if (!isVideoReady) {
            LoadingOverlay()
        }

        IconButton(
            onClick = onSettingsClick,
            modifier = Modifier
                .align(Alignment.TopEnd)
                .padding(8.dp)
        ) {
            Icon(Icons.Filled.Settings, "Quality Settings", tint = Color.White)
        }

        if (isVideoReady) {
            val badgeColor = if (isP2PActive) {
                MaterialTheme.colorScheme.primary
            } else {
                MaterialTheme.colorScheme.error
            }
            val badgeText = if (isP2PActive) "P2P ON" else "P2P OFF"
            val textColor = if (isP2PActive) {
                MaterialTheme.colorScheme.onPrimary
            } else {
                MaterialTheme.colorScheme.onError
            }

            Text(
                text = badgeText,
                color = textColor,
                style = MaterialTheme.typography.labelSmall,
                fontWeight = FontWeight.Bold,
                modifier = Modifier
                    .align(Alignment.TopStart)
                    .padding(8.dp)
                    .background(badgeColor.copy(alpha = 0.8f), RoundedCornerShape(4.dp))
                    .padding(horizontal = 6.dp, vertical = 4.dp)
            )
        }
    }
}

@Composable
private fun LoadingOverlay() {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(Color.Black),
        contentAlignment = Alignment.Center
    ) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            CircularProgressIndicator(
                color = MaterialTheme.colorScheme.primary,
                modifier = Modifier.size(48.dp)
            )

            Spacer(modifier = Modifier.height(8.dp))

            Text(
                text = "Loading Stream...",
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                style = MaterialTheme.typography.bodySmall
            )
        }
    }
}

File: iosDemo/iosDemo/UI/Screens/Player/Components/VideoPlayerSurface.swift
import SwiftUI
import AVKit

struct VideoPlayerSurface: View {
    let player: AVPlayer?
    let isP2PActive: Bool
    let isVideoReady: Bool
    let onSettingsClick: () -> Void

    var body: some View {
        ZStack {
            Color.black.aspectRatio(16/9, contentMode: .fit)

            if let player = player {
                VideoPlayer(player: player)
                    .aspectRatio(16/9, contentMode: .fit)
            }

            if !isVideoReady {
                VStack(spacing: 8) {
                    ProgressView().tint(AppTheme.primary)
                    Text("Loading Stream...")
                        .font(.caption)
                        .foregroundColor(AppTheme.onSurfaceVariant)
                }
            }

            VStack {
                HStack(alignment: .top) {
                    if isVideoReady {
                        Text(isP2PActive ? "P2P ON" : "P2P OFF")
                            .font(.caption2)
                            .fontWeight(.bold)
                            .padding(.horizontal, 6)
                            .padding(.vertical, 4)
                            .background(isP2PActive ? AppTheme.primary.opacity(0.8) : AppTheme.error.opacity(0.8))
                            .foregroundColor(isP2PActive ? AppTheme.onPrimary : AppTheme.onError)
                            .cornerRadius(4)
                            .padding(8)
                    }
                    Spacer()
                    Button(action: onSettingsClick) {
                        Image(systemName: "gearshape.fill")
                            .foregroundColor(.white)
                            .padding(8)
                    }
                }
                Spacer()
            }
        }
        .aspectRatio(16/9, contentMode: .fit)
    }
}
